enhancements/proposed/20200501-json-output/start-json-output.md

```diff
@@ -0,0 +1,231 @@
+
+## Reviewer Priorities
+
+Please review this proposal with the following priorities:
+
+*   Does this fit with minikube's [principles](https://minikube.sigs.k8s.io/docs/concepts/principles/)?
+*   Are there other approaches to consider?
+*   Could the implementation be made simpler?
+*   Are there usability, reliability, or technical debt concerns?
+
+Please leave the above text in your proposal as instructions to the reader.
+
+## Summary
+
+This proposal discusses adding JSON output to `minikube start`. 
+This feature will allow tools that rely on minikube, such as IDE extensions, to better parse errors and progress logs from STDOUT and STDERR on `minikube start`.
+This allows end users to see clear, and ideally actionable, error messages when minikube breaks.
+
+Minikube currently communicates state to users via:
+1. Logs via glog (sent to stderr on `minikube start`)
+1. Logs to stdout which represent a step on `minikube start` (e.g. "Preparing Kubernetes...")
+1. Logs to stdout which don't represent a step on `minikube start`. This can be an unexpected warning message or an `option`.
+
+This proposal focuses **only** on converting outputs representing clear steps in `minikube start` to JSON (option 2), and making sure error code sent to stderr is parsable.
+
+## Goals
+
+### Stderr
+*   Error code from [err_map.go](https://github.com/kubernetes/minikube/blob/master/pkg/minikube/problem/err_map.go) is parsable from stderr
+
+### Stdout
+*   Progress of each step of `minikube start` is communicated in JSON and includes:
+  1. A name for the current step (e.g. `Preparing Kubernetes`)
+  1. The number of the current step
+  1. The expected total number of steps
+*   Progress of artifacts as they are being downloaded is communicated in JSON
+
+
+
+## Non-Goals
+
+*   Change the way we handle logs via glog
+*   Change the way we handle non-steps to stdout (Warnings, etc.)
+*   Improving error handling in minikube; this is just about how we output errors to users
+
+## Expected Output
+
+Users can specify JSON output via a `--output json` flag. 
+
+### Stderr
+
+If minikube fails, and an actionable error message exists, the following JSON will be printed to stderr:
+
+```json
+{
+  "ID": "KVM_UNAVAILABLE",
+  "Err": {},
+  "Advice": "Your host does not support KVM virtualization. Ensure that qemu-kvm is installed, and run 'virt-host-validate' to debug the problem",
+  "URL": "http://mikko.repolainen.fi/documents/virtualization-with-kvm",
+  "Issues": [
+    2991
+  ],
+  "ShowIssueLink": false
+}
+```
+
+### Stdout
+If `--output json` is specified, each step of `minikube start` will be output as JSON:
+
+```json
+{
+  "Name": "Selecting Driver",
+  "Message": "✨  Using the hyperkit driver based on user configuration\n",
+  "TotalSteps": 9,
+  "CurrentStep": 2,
+  "Type": "Log"
+}
+```
+
+and each update on a downloaded artifact will be output as:
+
+```json
+{
+  "Type": "Download",
+  "Artifact": "preload.tar.gz",
+  "Progress": "10%"
```

Comment:
Consider `"CurrentStep": 2.1` if we're 10% of the way done with step 2. It conveys the same information, but readers might be able to do no processing and have their progress bars Just Work.